handlers/public.go

```diff
@@ -0,0 +1,208 @@
+	r.Get("/healthcheck", getHealthCheck).Name("get-healthcheck")
+	r.Post("/payments", createPaymentSession).Name("create-payment")
+}
+
+func getHealthCheck(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+var client http.Client
+
+func createPaymentSession(w http.ResponseWriter, req *http.Request) {
+	if req.Body == nil {
+		log.ErrorR(req, fmt.Errorf("request body empty"))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	requestDecoder := json.NewDecoder(req.Body)
+	var b createPaymentResource
+	err := requestDecoder.Decode(&b)
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("request body invalid"))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	resource := b.Resource
+
+	cfg, err := config.Get()
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("error getting config: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	url, err := url.Parse(resource)
+	if err != nil {
+		log.ErrorR(req, fmt.Errorf("error parsing resource: [%s]", err))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	whitelist := strings.Split(cfg.DomainWhitelist, ",")
+	matched := false
+	for _, domain := range whitelist {
+		if url.Host == domain {
+			matched = true
+			break
+		}
+	}
+	if !matched {
+		log.ErrorR(req, fmt.Errorf("invalid resource domain: %s", url.Host))
+		w.WriteHeader(http.StatusBadRequest) // 400
+		return
+	}
+
+	resourceReq, err := http.NewRequest("GET", resource, nil)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("failed to create Resource Request: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	resp, err := client.Do(resourceReq)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("error getting Cost Resource: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("error reading Cost Resource: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+
+	// TODO save cost resource and ensure all mandatory fields are present:
+
+	paymentResource := &PaymentResource{}
+	err = json.Unmarshal(body, paymentResource)
+	if err != nil {
+		log.ErrorR(resourceReq, fmt.Errorf("error reading Cost Resource: [%s]", err))
+		w.WriteHeader(http.StatusInternalServerError) // 500
+		return
+	}
+
+	user := strings.Split(req.Header.Get("Eric-Authorised-User"), ";")
+	email := user[0]
+	var forename string
+	var surname string
+
+	for i := 1; i < len(user); i++ {
+		v := strings.Split(user[i], "=")
+		if v[0] == " forename" {
+			forename = v[1]
+		} else if v[0] == " surname" {
+			surname = v[1]
+		} else {
+			log.ErrorR(req, fmt.Errorf("unexpected format in Eric-Authorised-User: %s", user))
+			w.WriteHeader(http.StatusInternalServerError) // 500
+			return
+		}
+	}
+	paymentResource.CreatedBy = CreatedBy{
```

Comment:
You could always have a new `func` that creates the `CreatedBy` data set, saves adding to the already long method.